objects/adffc0/contained/706d9f/contained/4363a8/script.lua

```diff
@@ -0,0 +1,299 @@
+    for _,guid in pairs(drowningTiles) do
+        local obj = getObjectFromGUID(guid)
+        if obj ~= nil then
+            obj.destruct()
+        end
+    end
+end
+
+-- Starts the Cast Down timer.
+function startCastDownTimer()
+    -- We store the reference to the Cast Down timer in Global, so every Deeps token has access to it.
+    -- Every Deeps token will try to do this, so if it already exists, don't worry about it.
+    if Global.getVar(timerVarName) then
+        return
+    end
+    -- Cast Down lasts 0.5 seconds.
+    -- We need to drop the tiles early, to give things time to fall, so do this every 0.2 seconds.
+    Wait.time(castDownCallback, 0.2, -1)
+    Global.setVar(timerVarName, true)
+end
+
+-- Casts down from the provided location, returning the board found.
+function getBoard(position)
+    local hits = Physics.cast({
+        origin = position, -- + Vector(0,0.2,0) [this was used temporarily while the models were slightly off]
+        direction = Vector(0,-1,0),
+        max_distance = 1,
+    })
+    for _,hit in pairs(hits) do
+        if Global.call("isIslandBoard", {obj=hit.hit_object}) then
+            return hit.hit_object
+        end
+    end
+    return nil
+end
+
+-- Marks a board as having been populated with drowned land tiles.
+-- Does so in a way that persists through saves and reloads.
+function markBoard(board, drowningTiles)
+    local json
+    if board.script_state ~= "" then
+        json = JSON.decode(board.script_state)
+    else
+        json = {}
+    end
+    json.drowningTiles = drowningTiles
+    board.script_state = JSON.encode(json)
+end
+
+-- Returns whether a board has been marked by markBoard.
+function checkBoard(board)
+    if board.script_state == "" then
+        return false
+    end
+    local json = JSON.decode(board.script_state)
+    return json.drowningTiles ~= nil
+end
+
+-- Gets the table of drowning tiles from a board
+function getBoardTiles(board)
+    if board.script_state == "" then
+        return {}
+    end
+    local json = JSON.decode(board.script_state)
+    return json.drowningTiles
+end
+
+-- Places a single drowned land tile, from a bag, into the correct place.
+-- Creates a copy from the bag, 'cause we may need multiple copies for extra bags.
+-- Gives the object a tag to indicate it's submerged in a board.
+function populateTile(bag, guid, position, rotation)
+    -- TODO: Scale this properly according to scaleFactors[SetupChecker.getVar("optionalScaleBoard")].size
+    -- Get the object out of the bag to clone it, but use a callback to put it back as soon as it spawns.
+    local original = bag.takeObject({
+        guid = guid,
+        position = bag.getPosition() + Vector(0, 10, 0),
+        callback_function = function(obj) bag.putObject(obj) end,
+    })
+    -- Spawn the clones somewhere out of the way, so we can lock them before moving them into position.
+    local clone = original.clone({position = original.getPosition() + Vector(0, 10, 0)})
```

Comment:
cloning is fine for now, but lets add a TODO to switch this over to using `spawnObjectData` instead so that way we don't need to pull the object out of bag and put it back in

Reply:
I wasn't under the impression that was a possibility. I don't know of any way to get the details of an object inside a bag except for the `getObjects()` function, and that doesn't contain nearly enough information about the stored objects to recreate them using `spawnObjectData()` (https://api.tabletopsimulator.com/object/#getobjects-containers).

Reply:
Oh, I see how this is supposed to happen; we can use `getData()`. I'll work on it.

Reply:
yeah it's a little tricky you need to get the data on the bag itself and then grab the specific contained object. there's a few places in the main mod that does it like this, namely setting up of the player area stuff. i grab data from the bag and set the player tints and then spawn in the object